pytorch_lightning/trainer/trainer.py

```diff
@@ -168,6 +172,18 @@ def __init__(
 
                     trainer = Trainer(early_stop_callback=early_stop_callback)
 
+            callback (:class:`.Callback`): Add a list of callbacks.
+                Example::
+                    from pytorch_lightning.callbacks import Callback
+                    class PrintCallback(Callback):
```

Comment:
callback should be passed in trainer and the model no?

Reply:
i mean, each call should have those args.

Reply:
Don't you prefer to have those variables as a callback property such as:

```python 
class Callback(abc.ABC):
    """Abstract base class used to build new callbacks."""

    def __init__(self):
        self._trainer = None

    @property
    def trainer(self):
        assert self._trainer is not None, _NO_TRAINER_ERROR_MSG
        return self._trainer

    @property
    def module(self):
        return self._trainer.get_model()

    @property
    def default_save_path(self):
        """Trainer default save path.
        """
        return self._trainer.default_save_path

    @property
    def rank(self):
        """Current trainer rank.
        """
        return self._trainer.proc_rank
```

It makes the code in `pl.Trainer` easier to read and update if needed. and overall make the callback class less verbose (especially since all callbacks have the same arguments).

Reply:
i’m not sure i agree. 

i thought about having a self.trainer and self.model but that means the user has to read the docs for this. whereas if you pass it in as an arg, it’s clear what’s available. 

but i see the value in both, so could go either way

Reply:
Let's wait for other comments. I don't have a strong feeling about it but I find the property pattern more elegant.

As for the doc, we should except the user to read it before using callbacks. This is not an extremely complicated pattern neither with plenty of parameters, so I except it should be fine.

Reply:
i disagree. “expecting” the user to read the docs means the api is not intuitive or obvious.

Reply:
I agree with @williamFalcon on this, arguments to the mthods seems much more intuitive

Reply:
Done.